### PR TITLE
KMeans implementation using Spark ML

### PR DESCRIPTION
**Pull request by blastarr**

### CODE REVIEW

File: analytics/javaspark/src/main/java/mil/nga/giat/geowave/analytic/javaspark/KMeansRunner.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+		// Retrieve the input centroids
+		centroidVectors = GeoWaveRDD.rddFeatureVectors(
+				javaPairRdd,
+				timeField,
+				scaledTimeRange);
+		centroidVectors.cache();
+
+		// Init the algorithm
+		KMeans kmeans = new KMeans();
+		kmeans.setInitializationMode("kmeans||");
```
Comment (blastarr) [inline analytics/javaspark/src/main/java/mil/nga/giat/geowave/analytic/javaspark/KMeansRunner.java]: reminder - create a new issue for this

File: analytics/javaspark/src/main/java/mil/nga/giat/geowave/analytic/javaspark/KMeansRunner.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+	public void setHost(
+			String host ) {
+		this.host = host;
+	}
+
+	public void setBoundingBox(
+			String bboxStr ) {
+		try {
+			// Expecting bbox in "LL-Lat LL-Lon UR-Lat UR-Lon" format
+			StringTokenizer tok = new StringTokenizer(
```
Comment (blastarr) [inline analytics/javaspark/src/main/java/mil/nga/giat/geowave/analytic/javaspark/KMeansRunner.java]: agree. we'll change it to x1 y1 x2 y2

File: analytics/javaspark/src/main/java/mil/nga/giat/geowave/analytic/javaspark/KMeansRunner.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+		if (inputDataStore == null) {
+			LOGGER.error("You must supply an input datastore!");
+			throw new IOException(
+					"You must supply an input datastore!");
+		}
+
+		QueryOptions queryOptions = null;
+		if (adapterId != null) {
+			// Retrieve the adapters
+			CloseableIterator<DataAdapter<?>> adapterIt = inputDataStore.createAdapterStore().getAdapters();
```
Comment (rfecher) [inline analytics/javaspark/src/main/java/mil/nga/giat/geowave/analytic/javaspark/KMeansRunner.java]: you can just setAdapterId() on queryoptions and not need to iterate through adapters to find a match

File: analytics/javaspark/src/test/java/mil/nga/giat/geowave/analytic/javaspark/TestSpark.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
+package mil.nga.giat.geowave.analytic.javaspark;
```
Comment (blastarr) [inline analytics/javaspark/src/test/java/mil/nga/giat/geowave/analytic/javaspark/TestSpark.java]: DELETE ME

File: extensions/adapters/vector/src/main/java/mil/nga/giat/geowave/adapter/vector/util/FeatureDataUtils.java [hunk 5]
```diff
@@ -221,4 +229,39 @@ public static SimpleFeature buildFeature(
...
 
+	public static String getFirstTimeField(
+			final DataStorePluginOptions dataStore,
+			final ByteArrayId adapterId ) {
+		AdapterStore adapterStore = dataStore.createAdapterStore();
+
+		DataAdapter adapter = adapterStore.getAdapter(adapterId);
+
+		if (adapter != null && adapter instanceof GeotoolsFeatureDataAdapter) {
+			GeotoolsFeatureDataAdapter gtAdapter = (GeotoolsFeatureDataAdapter) adapter;
```
Comment (rfecher) [inline extensions/adapters/vector/src/main/java/mil/nga/giat/geowave/adapter/vector/util/FeatureDataUtils.java]: adapter.getTimeDescriptors()

File: extensions/adapters/vector/src/main/java/mil/nga/giat/geowave/adapter/vector/utils/DateUtilities.java [hunk 6]
```diff
@@ -47,4 +55,21 @@ public static Date parseISO(
...
+
+	public static TemporalRange getTemporalRange(
+			final DataStorePluginOptions dataStorePlugin,
+			final ByteArrayId adapterId ) {
+		final DataStatisticsStore statisticsStore = dataStorePlugin.createDataStatisticsStore();
+		final CloseableIterator<DataStatistics<?>> statsIt = statisticsStore.getDataStatistics(adapterId);
+
+		while (statsIt.hasNext()) {
+			final DataStatistics stats = statsIt.next();
+			if (stats instanceof FeatureTimeRangeStatistics) {
```
Comment (rfecher) [inline extensions/adapters/vector/src/main/java/mil/nga/giat/geowave/adapter/vector/utils/DateUtilities.java]: FeatureTimeRangeStatistics.composeId(<fieldName>) gets you the statistics ID that can then be used within that call to DataStatisticsStore

**Pull request merged**